packages/compiler/src/core/checker.ts

```diff
@@ -560,15 +559,29 @@ export function createChecker(program: Program, resolver: NameResolver): Checker
     } else {
       entity = initial;
     }
-    // if (options.legacyTupleAndModelCast && entity !== null && isType(entity)) {
-    //   entity = legacy_tryTypeToValueCast(entity, constraint, node);
-    // }
     if (entity === null) {
       return null;
     }
     if (isValue(entity)) {
       return constraint ? inferScalarsFromConstraints(entity, constraint.type) : entity;
     }
+    // If a template parameter that can be a value is used in a template declaration then we allow it but we return null because we don't have an actual value.
+    if (
+      entity.kind === "TemplateParameter" &&
+      entity.constraint?.valueType &&
+      entity.constraint.type === undefined &&
+      mapper === undefined
+    ) {
+      return createValue(
+        {
+          entityKind: "Value",
+          valueKind: "NullValue",
```

Comment:
This kinda cheat quite a bit here. We are saying the value is `null` but it has a type of the constraint.

This value would really only be exposed if digging into a template declaration which isn't something you commonly do.